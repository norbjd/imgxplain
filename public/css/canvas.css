#canvases {
  position: relative;
}

#canvases canvas {
  position: absolute;
  top: 0;
  border: solid black;
}

#draw {
  z-index: 1;
}

#preview {
  z-index: 2;
}
