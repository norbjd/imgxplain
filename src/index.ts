import { App } from "./app";

window.addEventListener("DOMContentLoaded", () => {
  if (!window.AudioContext) {
    const errorMessage = "Web Audio not supported, change your browser";
    alert(errorMessage);
    throw new Error(errorMessage);
  }

  new App().run();
});
