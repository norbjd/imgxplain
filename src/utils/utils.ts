class Utils {
  static isEquivalent(a: unknown, b: unknown): boolean {
    if (typeof a != typeof b) {
      return false;
    }

    const aProps = Object.getOwnPropertyNames(a);
    const bProps = Object.getOwnPropertyNames(b);

    if (aProps.length != bProps.length) {
      return false;
    }

    for (let i = 0; i < aProps.length; i++) {
      const propName = aProps[i];
      if (a[propName] !== b[propName]) {
        return false;
      }
    }

    return true;
  }
}

export { Utils };
