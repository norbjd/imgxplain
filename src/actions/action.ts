import { Canvas } from "../canvas";

class Action<T extends ActionOptions> {
  canvas: Canvas;
  type: string;
  options: T;

  constructor(canvas: Canvas, type: string, options: T) {
    this.canvas = canvas;
    this.type = type;
    this.options = options;
  }

  do(context: CanvasRenderingContext2D): void {
    throw new Error("Not implemented");
  }

  execute(): void { this.do(this.canvas.mainContext); }
  preview(): void { this.do(this.canvas.previewContext); }
}

class ActionOptions {}

export {
  Action,
  ActionOptions,
};
