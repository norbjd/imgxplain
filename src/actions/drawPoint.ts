import { Action, ActionOptions } from "./action";
import { Canvas } from "../canvas";

class DrawPointOptions extends ActionOptions {
  centerX: number;
  centerY: number;
  radius: number;
  color: string;

  constructor(centerX: number, centerY: number, radius: number, color: string) {
    super();
    this.centerX = centerX;
    this.centerY = centerY;
    this.radius = radius;
    this.color = color;
  }
}

class DrawPoint extends Action<DrawPointOptions> {
  constructor(canvas: Canvas, options: DrawPointOptions) {
    super(canvas, "draw_point", options);
  }

  do(context: CanvasRenderingContext2D): void {
    const pointCanvasCenter = this.canvas.getPointProjectionOnCanvas(
      this.options.centerX,
      this.options.centerY
    );
    const pointCanvasCenterX = pointCanvasCenter.x;
    const pointCanvasCenterY = pointCanvasCenter.y;

    const ratio = this.canvas.getImageRatio();
    const pointCanvasRadiusX = this.options.radius / ratio.ratioWidth;
    const pointCanvasRadiusY = this.options.radius / ratio.ratioHeight;

    context.save();

    context.beginPath();

    context.fillStyle = this.options.color;
    context.strokeStyle = this.options.color;
    context.arc(
      pointCanvasCenterX,
      pointCanvasCenterY,
      pointCanvasRadiusX,
      0,
      Math.PI
    );
    context.fill();
    context.beginPath();
    context.arc(
      pointCanvasCenterX,
      pointCanvasCenterY,
      pointCanvasRadiusY,
      Math.PI,
      2 * Math.PI
    );
    context.fill();

    context.restore();
  }
}

export { DrawPoint, DrawPointOptions };
