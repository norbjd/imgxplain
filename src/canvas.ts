import { Action, ActionOptions } from "./actions/action";
import { Zoom, ZoomOptions } from "./actions/zoom";
import { DOMUtils } from "./utils/domUtils";

class Canvas {
  img: HTMLImageElement;
  imgTopLeftX: number;
  imgTopLeftY: number;
  imgBottomRightX: number;
  imgBottomRightY: number;

  mainCanvas: HTMLCanvasElement;
  mainContext: CanvasRenderingContext2D;

  previewCanvas: HTMLCanvasElement;
  previewContext: CanvasRenderingContext2D;

  clickPosition: { x: number; y: number };

  ccEnabled: boolean;

  constructor(imgBlob: string) {
    this.img = new Image();
    this.img.src = imgBlob;
    this.img.addEventListener("load", () => {
      this.renderImage();
    });

    this.imgTopLeftX = 0;
    this.imgTopLeftY = 0;
    this.imgBottomRightX = this.img.naturalWidth;
    this.imgBottomRightY = this.img.naturalHeight;

    this.mainCanvas = DOMUtils.getMainCanvas();
    this.mainContext = this.mainCanvas.getContext("2d");

    this.previewCanvas = DOMUtils.getPreviewCanvas();
    this.previewContext = this.previewCanvas.getContext("2d");

    this.clickPosition = { x: 0, y: 0 };

    this.ccEnabled = false;
  }

  render(
    topLeftX: number,
    topLeftY: number,
    bottomRightX: number,
    bottomRightY: number
  ): void {
    this.imgTopLeftX = topLeftX;
    this.imgTopLeftY = topLeftY;
    this.imgBottomRightX = bottomRightX;
    this.imgBottomRightY = bottomRightY;

    const imgWidth = this.imgBottomRightX - this.imgTopLeftX;
    const imgHeight = this.imgBottomRightY - this.imgTopLeftY;

    const imgRatio = imgWidth / imgHeight;

    const mainDivMaxWidth = DOMUtils.getAvailableMaxWidthForMainDiv();
    const mainDivMaxHeight = DOMUtils.getAvailableMaxHeightForMainDiv();

    let canvasWidth = Math.min(mainDivMaxWidth, imgWidth);
    let canvasHeight = canvasWidth / imgRatio;

    if (canvasHeight >= mainDivMaxHeight) {
      canvasHeight = mainDivMaxHeight;
      canvasWidth = canvasHeight * imgRatio;
    }

    this.mainCanvas.width = canvasWidth;
    this.mainCanvas.height = canvasHeight;
    this.mainContext.drawImage(
      this.img,
      this.imgTopLeftX,
      this.imgTopLeftY,
      imgWidth,
      imgHeight,
      0,
      0,
      canvasWidth,
      canvasHeight
    );

    this.previewCanvas.width = canvasWidth;
    this.previewCanvas.height = canvasHeight;
    this.previewCanvas.style.display = "none";

    // TODO : move somewhere else
    // -20 : padding
    DOMUtils.getTranscriptDiv().style.width = "" + (canvasWidth + 5 - 20);
  }

  renderFullImage(): void {
    this.render(0, 0, this.img.naturalWidth, this.img.naturalHeight);
  }

  renderImage(): void {
    this.render(
      this.imgTopLeftX,
      this.imgTopLeftY,
      this.imgBottomRightX,
      this.imgBottomRightY
    );
  }

  getWidthAndHeight(): { width: number; height: number } {
    return {
      width: this.mainCanvas.width,
      height: this.mainCanvas.height,
    };
  }

  getImageRatio(): { ratioWidth: number; ratioHeight: number } {
    const imgWidth = this.imgBottomRightX - this.imgTopLeftX;
    const imgHeight = this.imgBottomRightY - this.imgTopLeftY;

    const canvasWidthAndHeight = this.getWidthAndHeight();

    return {
      ratioWidth: imgWidth / canvasWidthAndHeight.width,
      ratioHeight: imgHeight / canvasWidthAndHeight.height,
    };
  }

  getPointProjectionOnCanvas(
    imgPointX: number,
    imgPointY: number
  ): { x: number; y: number } {
    const ratio = this.getImageRatio();

    const canvasPointX = (imgPointX - this.imgTopLeftX) / ratio.ratioWidth;
    const canvasPointY = (imgPointY - this.imgTopLeftY) / ratio.ratioHeight;

    return {
      x: canvasPointX,
      y: canvasPointY,
    };
  }

  getCursorPositionOnCanvas(event: MouseEvent): { x: number; y: number } {
    const rect = this.mainCanvas.getBoundingClientRect();
    const x = event.clientX - rect.left;
    const y = event.clientY - rect.top;

    const ratio = this.getImageRatio();

    return {
      x: Math.floor(x * ratio.ratioWidth) + this.imgTopLeftX,
      y: Math.floor(y * ratio.ratioHeight) + this.imgTopLeftY,
    };
  }

  addClickEventListener(listener: (event: MouseEvent) => void): void {
    this.mainCanvas.addEventListener("click", listener);
  }

  executeAction(action: Action<ActionOptions>): void {
    action.execute();
  }

  executeActions(actions: Action<ActionOptions>[]): void {
    if (actions.find((action) => action instanceof Zoom) == undefined) {
      this.renderFullImage();
    }

    this.renderImage();
    actions.forEach((action) => {
      this.executeAction(action);
    });
  }

  previewAction(action: Action<ActionOptions>, timeInMs: number): void {
    this.previewCanvas.style.display = "block";
    action.preview();
    setTimeout(() => {
      this.previewContext.clearRect(
        0,
        0,
        this.previewCanvas.width,
        this.previewCanvas.height
      );
      this.previewCanvas.style.display = "none";
      if (action instanceof Zoom) {
        const unzoom = new Zoom(
          this,
          new ZoomOptions(0, 0, this.img.naturalWidth, this.img.naturalHeight)
        );
        this.executeAction(unzoom);
      }
    }, timeInMs);
  }
}

export { Canvas };
