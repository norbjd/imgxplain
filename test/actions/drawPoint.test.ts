import { expect, it, jest, beforeEach } from '@jest/globals';
import { DrawPoint, DrawPointOptions } from "../../src/actions/drawPoint";
import { Canvas } from '../../src/canvas';
import { DOMUtils } from '../../src/utils/domUtils';

import { readFileSync } from "fs";
import { resolve } from "path";
import { initCanvas } from './utils';

const html = readFileSync(resolve(__dirname, '../../public/index.html'), 'utf8');
jest.dontMock('fs');
document.documentElement.innerHTML = html.toString();

let canvas: Canvas;
beforeEach(() => {
    canvas = initCanvas();
});

it("should draw a red point in the middle", () => {
    new DrawPoint(canvas, new DrawPointOptions(
        150, 75, 50, "#ff0000"
    )).execute();
    const imgBase64 = DOMUtils.getMainCanvas().toDataURL("image/png");
    expect(imgBase64).toMatchSnapshot();
});

it("should draw a blue point on the top left", () => {
    new DrawPoint(canvas, new DrawPointOptions(
        0, 0, 50, "#0000ff"
    )).execute();
    const imgBase64 = DOMUtils.getMainCanvas().toDataURL("image/png");
    expect(imgBase64).toMatchSnapshot();
});

it("should preview a red point in the middle", () => {
    new DrawPoint(canvas, new DrawPointOptions(
        150, 75, 50, "#ff0000"
    )).preview();
    const imgBase64 = DOMUtils.getPreviewCanvas().toDataURL("image/png");
    expect(imgBase64).toMatchSnapshot();
});

it("should preview a blue point on the top left", () => {
    new DrawPoint(canvas, new DrawPointOptions(
        0, 0, 50, "#0000ff"
    )).preview();
    const imgBase64 = DOMUtils.getPreviewCanvas().toDataURL("image/png");
    expect(imgBase64).toMatchSnapshot();
});
