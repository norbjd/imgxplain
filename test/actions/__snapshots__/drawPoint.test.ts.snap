// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should draw a blue point on the top left 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAEKElEQVR4nO3dPWsUaxzG4f8eTBbEJCYicX0B30LUSjsNNqIErCwMaGEUG0sttPVL2B/EDyEiFtoIVjY2WtgIfge7nGIrUTgiyTxzZ64LAql27urHM5vZzahqa6vCjcdVP360XgHstH9aD9gOc3OtFwBdECwghmABMQQLiLErgrW83HoB0IVdEawzZ1ovALqwK4K1utp6AdAFwQJi7Ipgray0XgB0IT5YJ05ULS21XgF0IT5YV6+2XgB0JT5YV660XgB0ZZT84efRqOr796rJpPUSoAvRJ6yzZ8UKhiQ6WBsbrRcAXYq+Jfz82TNYMCSxJ6xLl8QKhiY2WJubrRcAXYu8Jdy7t+rbt6oDB1ovAboUecJ68ECsYIjiTljjcdXXr1VHjrReAnQt7oR1/75YwVBFnbBmZqq+fJl+4BkYnqgT1qNHYgVDFnPCmkymD4rOz7deArQSc8J69kysYOgiTljr61WvX7deAbTW+2AtLlZ9/Fh1/HjrJUBrvb4lHI2qnj8XK2Cq18F68qTqxo3WK4C+6O0t4dpa1bt302evAKp6GqxTp6rev/cv6IGf9e6WcDKpevNGrIBf9SpY8/NVL196mh34vd4Ea2FhGqsLF1ovAfpqT+sBVVWHDlW9elV1/nzrJUCfNQ/WyZPTp9hPn269BOi7preEa2tVHz6IFfBnmgRrNKp6+LDq7duqgwdbLAASdX5LuLBQ9e+/VTdvdn1lIF2nJ6z19apPn8QK+DudBOvw4aoXL6Zvrh871sUVgd1oR4M1M1P1+PH0e9jv3t3JKwFDsCPvYc3OVt26VfX0adXKyk5cARiibQ3WeFx17940VEePbucrA2xTsC5erNrcrLp9u2ppaTteEeBXfx2sc+eqNjaq7txx2wd044+DNZlUXb5cde1a1fXr/toHdG/P7GzVvn1V+/dXzc1Nf19erlpd/flncbH1VGDoRltb/fvGUYDf6c33YQH8H8ECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEOM/ms5j00UoBasAAAAASUVORK5CYII="`;

exports[`should draw a red point in the middle 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAJjklEQVR4nO3dfWiV5R/H8c+JrVG6rW3Kdmy21nRzQtSKskR7wGFJ2eNAA5dG4R8RSVRQZET0j0JP1H+SVI7MCAOzB8QNtQwSeoCM9iBzMTc7cw9aOmdzsd8fF4N+c+q2tn2v677fLxjuD9k+59ycz77XOdd934nBwcFBAUAALrEOAACjRWEBCAaFBSAYFBaAYFBYAIJBYQEIBoUFIBgUFoBgUFgAgkFhAQgGhQUgGBQWgGBQWACCQWEBCAaFBSAYFBaAYFBYAIJBYQEIBoUFIBgUFoBgUFgAgkFhAQgGhQUgGBQWgGCkWQdAYLq7pcZGqaFBampy36dSUm+v+zp+3P0rSdOmSTk57t9p06RkUiotdV/z5kllZVJenu3jQVAS3PkZF5RKSd9+K9XWSrt3Sy0tE/vzk0lp0SKpslK66y6pqGhifz4ihcLCuerrpS1bpO3bpUOHpvZ3l5ZKVVVSdbWbwoB/obDgdHVJW7dKNTXSDz9Yp3FuvllatUp65BFpxgzrNPAAhRV3ra3SG29I770nnT5tnWZkGRnS6tXS+vXS7NnWaWCIwoqrlhbp7belTZukM2es04xOerq0cqX00kvuDXvEDoUVN7290muvualqYMA6zfikpUlPPukeR1aWdRpMIQorTnbudC/0tjbrJBMjmZQ2bHBv0CcS1mkwBSisOGhpkR5/XNqzxzrJ5FiyRNq8mS0RMcBO96jbsUO68cbolpUk1dVJ110nffqpdRJMMgorqs6elV54QXrwQbf7POr+/FNasUJat07q77dOg0nCkjCK/vhDuu8+f/ZTTbUFC9xkmZ9vnQQTjMKKmpYWd4rLVO9Q901xsbRrlzR3rnUSTCCWhFFy8KA7Ly/uZSW54l68WPr5Z+skmEAUVlTs2+fK6uhR6yT+6OiQ7rjDnbyNSGBJGAW//CLdfrt04oR1Ej9lZblPSW+4wToJ/iMKK3TNzW6ySqWsk/ht5kxp/353NQgEiyVhyI4dk5Yto6xGo7OT5yoCKKxQ/f23dM89vME+FocPSw88wD6tgFFYoXr++fjus/ovDhyQXnzROgXGifewQvTFF25jKIdufBIJ6bPP3LSFoFBYoWltlSoqpJ4e6yRhy8mRfvzRbTBFMFgShmRwUFqzhrKaCMePS2vXWqfAGFFYIdm6NdpXXZhqtbXSJ59Yp8AYsCQMxV9/SeXl7GSfaAUF7h6L2dnWSTAKTFihWL+espoMqZT06qvWKTBKTFghOHTITVf//GOdJJrS0tyUVVJinQQXwYQVgg0bKKvJNDAgbdxonQKjwITluyNHpDlz2J092dLT3STLdeG9xoTlu40bKaupcPas9Oab1ilwEUxYPuvsdH/x+/qsk8TDZZe5iTYvzzoJzoMJy2cff0xZTaW+PmnbNusUuAAKy2c1NdYJ4ofn3GssCX1VXy/Nn2+dIp4aGqSyMusUGAETlq+2bLFOEF8ffWSdAOfBhOWr0lIuzmelvFz67TfrFBgBheWj9napsNA6Rby1tUlXXmmdAsOwJPRRXZ11Auzda50AI6CwfMQlZOxxDLzEktBHRUXuyqKwU1zsbloBr1BYvunqcvfQg73ubik31zoF/oUloW8aG60TYEhTk3UCDENh+YbC8gfHwjsUlm94kfiDY+EdCss3LEP8QWF5h8LyTSplnQBDOjqsE2AYCss3p05ZJ8CQkyetE2AYCss3vEj8wbHwDoXlG14k/uBYeIfC8g1LQn9QWN6hsAAEg8LyzfTp1gkwJDPTOgGGobB8w4vEHxwL71BYvuFF4g+OhXcoLN+wJPQHheUdCss3BQXWCTAkP986AYahsHzD7aX8MW+edQIMQ2H5hsLyB8fCOxSWb3iR+INj4R0ukeybnh4pL886BSQukewhJizf5Oa6m1DAVnExZeUhCstHd95pnQBLllgnwAgoLB9RWPY4Bl7iPSwfHT3KbdItJRJSe7uUTFonwTBMWD6aNYtPqCyVl1NWnqKwfPXQQ9YJ4quqyjoBzoMloa+ampiyrDQ08Nx7ignLV6Wl0k03WaeIn1tvpaw8RmH5rLraOkH88Jx7jSWhz7q6pKuukvr6rJPEw+WXS62tnGngMSYsn82YIT3xhHWK+Fi7lrLyHBOW79rapJISqb/fOkm0ZWRIzc3sf/McE5bvCgulRx+1ThF9jz1GWQWACSsEzc3uYnIDA9ZJoik9XWpsdCc8w2tMWCEoKZGeeso6RXStW0dZBYIJKxQnT7pTRtrbrZNESzLpNopmZVknwSgwYYUiM1N6/XXrFNHzzjuUVUCYsEJTWSnV1VmniIalS6Vdu6xTYAworNAcOSJVVLjL92L8cnKkn36Srr7aOgnGgCVhaGbPlj74wF2zCeOTSEjvv09ZBYjCCtG997pPtjA+zz0n3X+/dQqMA0vCUPX3S7fdJh04YJ0kLAsXSnv3ur1XCA6FFbLOTmnxYrfpERdXUiJ99x23oA8YS8KQzZwpffUVl/MdjWRS2r2bsgochRW6a65xH81fcYV1En9lZUlffslu9gigsKLg2mulHTuk7GzrJP7JznZlVVFhnQQTgPewouTXX6W77+b0nSEFBdLXX0vXX2+dBBOEwoqalhZXWk1N1klsDS2V58yxToIJxJIwaoqLpW++kRYssE5iZ+FC6fvvKasIorCiKD9f2r9feuUV6ZIYHeJEQnr6aWnPHvcJKiKHJWHU7dwprVkj9fRYJ5lc2dnS5s3Sww9bJ8EkitGf35havtyd5FtZaZ1k8ixdKh08SFnFAIUVB0VFbtPk55+7k6ejYtYs6cMP3ZvrUXpcOC8KK06WL3dbH555RkpLs04zfunp0rPPulOSuEFHrPAeVlz9/rv01lvSpk3SmTPWaUbn0kulFSukl1+W5s61TgMDFFbcdXS44nr3Xen0aes0I8vIkFavdkVVWGidBoYoLDjd3dK2bVJNjT+XrLnlFqm6Wlq5UsrNtU4DD1BYOFdTkyuu7dul+vqp/d3z50tVVdKqVSz7cA4KCxd27Ji0b59UW+u+Dh+e2J+fTEqLFrltF8uW8WkfLojCwtj09LgJrKHBfUrX1CSlUlJvr7t34okT0qlT7v9On+4ue5OZ6b7Pz5fKyv7/KyfH9vEgKBQWgGCwDwtAMCgsAMGgsAAEg8ICEAwKC0AwKCwAwaCwAASDwgIQDAoLQDAoLADBoLAABIPCAhAMCgtAMCgsAMGgsAAEg8ICEAwKC0AwKCwAwaCwAASDwgIQDAoLQDAoLADBoLAABIPCAhCM/wF8cahYcu62XQAAAABJRU5ErkJggg=="`;

exports[`should preview a blue point on the top left 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAEKElEQVR4nO3dPWsUaxzG4f8eTBbEJCYicX0B30LUSjsNNqIErCwMaGEUG0sttPVL2B/EDyEiFtoIVjY2WtgIfge7nGIrUTgiyTxzZ64LAql27urHM5vZzahqa6vCjcdVP360XgHstH9aD9gOc3OtFwBdECwghmABMQQLiLErgrW83HoB0IVdEawzZ1ovALqwK4K1utp6AdAFwQJi7Ipgray0XgB0IT5YJ05ULS21XgF0IT5YV6+2XgB0JT5YV660XgB0ZZT84efRqOr796rJpPUSoAvRJ6yzZ8UKhiQ6WBsbrRcAXYq+Jfz82TNYMCSxJ6xLl8QKhiY2WJubrRcAXYu8Jdy7t+rbt6oDB1ovAboUecJ68ECsYIjiTljjcdXXr1VHjrReAnQt7oR1/75YwVBFnbBmZqq+fJl+4BkYnqgT1qNHYgVDFnPCmkymD4rOz7deArQSc8J69kysYOgiTljr61WvX7deAbTW+2AtLlZ9/Fh1/HjrJUBrvb4lHI2qnj8XK2Cq18F68qTqxo3WK4C+6O0t4dpa1bt302evAKp6GqxTp6rev/cv6IGf9e6WcDKpevNGrIBf9SpY8/NVL196mh34vd4Ea2FhGqsLF1ovAfpqT+sBVVWHDlW9elV1/nzrJUCfNQ/WyZPTp9hPn269BOi7preEa2tVHz6IFfBnmgRrNKp6+LDq7duqgwdbLAASdX5LuLBQ9e+/VTdvdn1lIF2nJ6z19apPn8QK+DudBOvw4aoXL6Zvrh871sUVgd1oR4M1M1P1+PH0e9jv3t3JKwFDsCPvYc3OVt26VfX0adXKyk5cARiibQ3WeFx17940VEePbucrA2xTsC5erNrcrLp9u2ppaTteEeBXfx2sc+eqNjaq7txx2wd044+DNZlUXb5cde1a1fXr/toHdG/P7GzVvn1V+/dXzc1Nf19erlpd/flncbH1VGDoRltb/fvGUYDf6c33YQH8H8ECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEOM/ms5j00UoBasAAAAASUVORK5CYII="`;

exports[`should preview a red point in the middle 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAJjklEQVR4nO3dfWiV5R/H8c+JrVG6rW3Kdmy21nRzQtSKskR7wGFJ2eNAA5dG4R8RSVRQZET0j0JP1H+SVI7MCAOzB8QNtQwSeoCM9iBzMTc7cw9aOmdzsd8fF4N+c+q2tn2v677fLxjuD9k+59ycz77XOdd934nBwcFBAUAALrEOAACjRWEBCAaFBSAYFBaAYFBYAIJBYQEIBoUFIBgUFoBgUFgAgkFhAQgGhQUgGBQWgGBQWACCQWEBCAaFBSAYFBaAYFBYAIJBYQEIBoUFIBgUFoBgUFgAgkFhAQgGhQUgGBQWgGCkWQdAYLq7pcZGqaFBampy36dSUm+v+zp+3P0rSdOmSTk57t9p06RkUiotdV/z5kllZVJenu3jQVAS3PkZF5RKSd9+K9XWSrt3Sy0tE/vzk0lp0SKpslK66y6pqGhifz4ihcLCuerrpS1bpO3bpUOHpvZ3l5ZKVVVSdbWbwoB/obDgdHVJW7dKNTXSDz9Yp3FuvllatUp65BFpxgzrNPAAhRV3ra3SG29I770nnT5tnWZkGRnS6tXS+vXS7NnWaWCIwoqrlhbp7belTZukM2es04xOerq0cqX00kvuDXvEDoUVN7290muvualqYMA6zfikpUlPPukeR1aWdRpMIQorTnbudC/0tjbrJBMjmZQ2bHBv0CcS1mkwBSisOGhpkR5/XNqzxzrJ5FiyRNq8mS0RMcBO96jbsUO68cbolpUk1dVJ110nffqpdRJMMgorqs6elV54QXrwQbf7POr+/FNasUJat07q77dOg0nCkjCK/vhDuu8+f/ZTTbUFC9xkmZ9vnQQTjMKKmpYWd4rLVO9Q901xsbRrlzR3rnUSTCCWhFFy8KA7Ly/uZSW54l68WPr5Z+skmEAUVlTs2+fK6uhR6yT+6OiQ7rjDnbyNSGBJGAW//CLdfrt04oR1Ej9lZblPSW+4wToJ/iMKK3TNzW6ySqWsk/ht5kxp/353NQgEiyVhyI4dk5Yto6xGo7OT5yoCKKxQ/f23dM89vME+FocPSw88wD6tgFFYoXr++fjus/ovDhyQXnzROgXGifewQvTFF25jKIdufBIJ6bPP3LSFoFBYoWltlSoqpJ4e6yRhy8mRfvzRbTBFMFgShmRwUFqzhrKaCMePS2vXWqfAGFFYIdm6NdpXXZhqtbXSJ59Yp8AYsCQMxV9/SeXl7GSfaAUF7h6L2dnWSTAKTFihWL+espoMqZT06qvWKTBKTFghOHTITVf//GOdJJrS0tyUVVJinQQXwYQVgg0bKKvJNDAgbdxonQKjwITluyNHpDlz2J092dLT3STLdeG9xoTlu40bKaupcPas9Oab1ilwEUxYPuvsdH/x+/qsk8TDZZe5iTYvzzoJzoMJy2cff0xZTaW+PmnbNusUuAAKy2c1NdYJ4ofn3GssCX1VXy/Nn2+dIp4aGqSyMusUGAETlq+2bLFOEF8ffWSdAOfBhOWr0lIuzmelvFz67TfrFBgBheWj9napsNA6Rby1tUlXXmmdAsOwJPRRXZ11Auzda50AI6CwfMQlZOxxDLzEktBHRUXuyqKwU1zsbloBr1BYvunqcvfQg73ubik31zoF/oUloW8aG60TYEhTk3UCDENh+YbC8gfHwjsUlm94kfiDY+EdCss3LEP8QWF5h8LyTSplnQBDOjqsE2AYCss3p05ZJ8CQkyetE2AYCss3vEj8wbHwDoXlG14k/uBYeIfC8g1LQn9QWN6hsAAEg8LyzfTp1gkwJDPTOgGGobB8w4vEHxwL71BYvuFF4g+OhXcoLN+wJPQHheUdCss3BQXWCTAkP986AYahsHzD7aX8MW+edQIMQ2H5hsLyB8fCOxSWb3iR+INj4R0ukeybnh4pL886BSQukewhJizf5Oa6m1DAVnExZeUhCstHd95pnQBLllgnwAgoLB9RWPY4Bl7iPSwfHT3KbdItJRJSe7uUTFonwTBMWD6aNYtPqCyVl1NWnqKwfPXQQ9YJ4quqyjoBzoMloa+ampiyrDQ08Nx7ignLV6Wl0k03WaeIn1tvpaw8RmH5rLraOkH88Jx7jSWhz7q6pKuukvr6rJPEw+WXS62tnGngMSYsn82YIT3xhHWK+Fi7lrLyHBOW79rapJISqb/fOkm0ZWRIzc3sf/McE5bvCgulRx+1ThF9jz1GWQWACSsEzc3uYnIDA9ZJoik9XWpsdCc8w2tMWCEoKZGeeso6RXStW0dZBYIJKxQnT7pTRtrbrZNESzLpNopmZVknwSgwYYUiM1N6/XXrFNHzzjuUVUCYsEJTWSnV1VmniIalS6Vdu6xTYAworNAcOSJVVLjL92L8cnKkn36Srr7aOgnGgCVhaGbPlj74wF2zCeOTSEjvv09ZBYjCCtG997pPtjA+zz0n3X+/dQqMA0vCUPX3S7fdJh04YJ0kLAsXSnv3ur1XCA6FFbLOTmnxYrfpERdXUiJ99x23oA8YS8KQzZwpffUVl/MdjWRS2r2bsgochRW6a65xH81fcYV1En9lZUlffslu9gigsKLg2mulHTuk7GzrJP7JznZlVVFhnQQTgPewouTXX6W77+b0nSEFBdLXX0vXX2+dBBOEwoqalhZXWk1N1klsDS2V58yxToIJxJIwaoqLpW++kRYssE5iZ+FC6fvvKasIorCiKD9f2r9feuUV6ZIYHeJEQnr6aWnPHvcJKiKHJWHU7dwprVkj9fRYJ5lc2dnS5s3Sww9bJ8EkitGf35havtyd5FtZaZ1k8ixdKh08SFnFAIUVB0VFbtPk55+7k6ejYtYs6cMP3ZvrUXpcOC8KK06WL3dbH555RkpLs04zfunp0rPPulOSuEFHrPAeVlz9/rv01lvSpk3SmTPWaUbn0kulFSukl1+W5s61TgMDFFbcdXS44nr3Xen0aes0I8vIkFavdkVVWGidBoYoLDjd3dK2bVJNjT+XrLnlFqm6Wlq5UsrNtU4DD1BYOFdTkyuu7dul+vqp/d3z50tVVdKqVSz7cA4KCxd27Ji0b59UW+u+Dh+e2J+fTEqLFrltF8uW8WkfLojCwtj09LgJrKHBfUrX1CSlUlJvr7t34okT0qlT7v9On+4ue5OZ6b7Pz5fKyv7/KyfH9vEgKBQWgGCwDwtAMCgsAMGgsAAEg8ICEAwKC0AwKCwAwaCwAASDwgIQDAoLQDAoLADBoLAABIPCAhAMCgtAMCgsAMGgsAAEg8ICEAwKC0AwKCwAwaCwAASDwgIQDAoLQDAoLADBoLAABIPCAhCM/wF8cahYcu62XQAAAABJRU5ErkJggg=="`;
